ui/src/saros/ui/browser/SWTBrowser.java

```diff
@@ -0,0 +1,200 @@
+package saros.ui.browser;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.log4j.Logger;
+import org.eclipse.swt.browser.Browser;
+import org.eclipse.swt.browser.BrowserFunction;
+import org.eclipse.swt.browser.ProgressEvent;
+import org.eclipse.swt.browser.ProgressListener;
+import org.eclipse.swt.events.DisposeEvent;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Display;
+
+public class SWTBrowser implements IBrowser {
+
+  private enum ReadyState {
+    LOADING,
+    COMPLETE
+  }
+
+  private static final Logger log = Logger.getLogger(SWTBrowser.class);
+
+  private final Browser browser;
+  private ReadyState state = ReadyState.LOADING;
+  private List<Runnable> runOnDisposalList =
+      Collections.synchronizedList(new ArrayList<Runnable>());
+
+  private final ProgressListener readyStateListener =
+      new ProgressListener() {
+        @Override
+        public void completed(ProgressEvent event) {
+          state = ReadyState.COMPLETE;
+        }
+
+        @Override
+        public void changed(ProgressEvent event) {
+          state = ReadyState.LOADING;
+        }
+      };
+
+  public SWTBrowser(Composite parent, int style) {
+    browser = new Browser(parent, style);
+
+    browser.addDisposeListener(
+        (DisposeEvent event) -> {
+          synchronized (runOnDisposalList) {
+            for (Runnable runnable : runOnDisposalList) {
+              try {
+                runnable.run();
+              } catch (RuntimeException exception) {
+                log.error("error running on disposal: " + exception.getMessage(), exception);
```

Comment:
I really do not find that kind of logging useful. You will find the message of the exception anyway.

E.G

0 [main] ERROR saros.activities.ActivityOptimizerTest  - error wile reading file: \foo (The system cannot find the file specified)
java.io.FileNotFoundException: \foo (The system cannot find the file specified)